### PR TITLE
Add test cases for serverExec

### LINKED ISSUE #90: Provide tests for `ServeExec`
# Summary

A test for `ServeExec` should be provided to cover at least: 
- that a http request with a not expected method fails
- that a request with a not existing token fails
- that a token can't be used twice
- that a request with a valid token and a not existing machine fails

This should be implemented using `net/http/httptest`. 

_Originally posted by @lukasfrank in https://github.com/ironcore-dev/libvirt-provider/pull/65#discussion_r1432477420_

### PR DESCRIPTION
**Pull request by ushabelgur**
# Proposed Changes

- add test  for validating http request with a not expected method fails
- add test for validating that a token can't be used twice
- add test to validate a request with a valid token and a not existing machine fails

Fixes #90

### CODE REVIEW

File: provider/server/exec_test.go [hunk 1]
```diff
@@ -68,5 +67,69 @@ var _ = Describe("Exec", func() {
...
 	})
 })
+
+func runExec(ctx context.Context, execUrl *url.URL) error {
+	randomSize := 1024 * 1024
+	randomData := make([]byte, randomSize)
+	var stdout bytes.Buffer
+
+	tty := term.TTY{
+		In:     bytes.NewReader(randomData),
```
Comment (lukas016) [inline provider/server/exec_test.go:109]: what are you testing here?

File: provider/server/exec_test.go [hunk 2]
```diff
@@ -68,5 +67,69 @@ var _ = Describe("Exec", func() {
...
+
+		By("issuing exec for non existing machine")
+		err = runExec(ctx, parsedResUrl)
+		machineNotSynchErr := fmt.Sprintf("machine %s has not yet been synced", createResp.Machine.Metadata.Id)
+		Expect(err).To(SatisfyAny(MatchError(ContainSubstring("404 page not found")), MatchError(ContainSubstring(machineNotSynchErr))))
+
 	})
 })
+
+func runExec(ctx context.Context, execUrl *url.URL) error {
```
Comment (lukas016) [inline provider/server/exec_test.go:103]: this function doesn't try send any command to console or read output. So you don't test whole logic of Streaming server. You test just http server logic but i recommend test whole streaming logic with send command and read output. Maybe login of user will enough.

**Review by lukas016 — CHANGES_REQUESTED** [review]
machine isn't deleted, when test fail

File: provider/server/exec_test.go [hunk 3]
```diff
@@ -29,15 +37,6 @@ var _ = Describe("Exec", func() {
 		Expect(err).NotTo(HaveOccurred())
 		Expect(createResp).NotTo(BeNil())
 
-		DeferCleanup(func(ctx SpecContext) {
```
[the remark this answers is not in the document]
Comment (ushabelgur) [reply provider/server/exec_test.go:32]: I moved delete to end of the test case to cover to this case: to validate a request with a valid token and a not existing machine fails

File: provider/server/exec_test.go (hunk 2 above)
Comment (ushabelgur) [reply provider/server/exec_test.go:103]: It was not mentioned in the scope of issue https://github.com/ironcore-dev/libvirt-provider/issues/90 If really i can cover. It talks about only negative cases.

File: provider/server/exec_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukas016) [reply provider/server/exec_test.go:32]: if test fail, machine will still exist what can affect another tests. So you need properly cleanup env after end of test and it doesn't matter on result of test.

File: provider/server/exec_test.go (hunk 2 above)
Comment (lukas016) [reply provider/server/exec_test.go:103]: we can discuss it on daily. But implement whole logic can help us catch error during development later.

File: provider/server/exec_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ushabelgur) [reply provider/server/exec_test.go:32]: Added back deletion part in DefereCleanup

**Review by hardikdr — COMMENTED** [review]
Thanks for the PR @ushabelgur .

Can you please inform here when the requested changes has already been taken care of.

**Comment by ushabelgur** [comment]
@hardikdr Machine deletion part at the end of test case has been addressed.  But the test case to validate whole streaming logic with send command and read output, I am still working on it.

File: provider/server/exec_test.go (hunk 2 above)
Comment (ushabelgur) [reply provider/server/exec_test.go:115]: We are using go routine with infinite loop to read and write to stream, even when i am sending Escape Sequence to terminate the stream it is not coming out of the go routine. Tried with bytes.Buffer Read and Write methods, somehow escape sequence is not working with this kind of buffer. I have created separate issue to work on it further #207

File: provider/server/exec_test.go [hunk 4]
```diff
@@ -30,9 +38,12 @@ var _ = Describe("Exec", func() {
 		Expect(createResp).NotTo(BeNil())
 
 		DeferCleanup(func(ctx SpecContext) {
-			_, err := machineClient.DeleteMachine(ctx, &iri.DeleteMachineRequest{MachineId: createResp.Machine.Metadata.Id})
-			Expect(err).ShouldNot(HaveOccurred())
 			Eventually(func() bool {
+				_, err := machineClient.DeleteMachine(ctx, &iri.DeleteMachineRequest{MachineId: createResp.Machine.Metadata.Id})
```
Comment (lukas016) [inline provider/server/exec_test.go:42]: pls call DeleteMachine just one time before Eventually

Comment (ushabelgur) [reply provider/server/exec_test.go:42]: We can not take it out of Eventually, we were facing this machine object getting overridden problem sometimes as mentioned in #164 . In all other integration test also its been handled similar way. May be after this fix we can tryout.

**Review by lukas016 — DISMISSED** [review]
LGTM, tests will improve in follow up issue: https://github.com/ironcore-dev/libvirt-provider/issues/207

**Review by lukas016 — APPROVED** [review]
LGTM

**Review by hardikdr — APPROVED** [review]
Thanks for the PR @ushabelgur, appreciate it.

**Pull request merged**